sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -939,16 +947,15 @@ public void processElement(ProcessContext c) throws Exception {
   }
 
   private static class WriteToSpannerFn
-      extends DoFn<Iterable<Mutation>, Void> {
-    private static final int MAX_RETRIES = 5;
-    private static final FluentBackoff BUNDLE_WRITE_BACKOFF = FluentBackoff.DEFAULT
-        .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+      extends DoFn<Iterable<MutationGroup>, MutationGroup> {
```

Comment:
Could you please make this a DoFn returning some success indicator on the main output, and failed mutations on an additional output? That way this will be compatible with the Wait transform https://github.com/apache/beam/pull/4301.

Reply:
What do you mean by a success indicator? Isn't returning empty PCollection of failed mutations good enough?

Reply:
I guess it's technically enough in this PR, but it can lead to issues in the future. E.g. in BigQueryIO, when using "batch loads", the collection of failed records is simply hard-coded as an empty collection, and it doesn't have a data dependency on writing successful records. If something like this is ever introduced to Spanner (e.g. bulk import), it can have a similar issue. I think, if a user wants to sequence something after a successful write, they should apply it to something that explicitly represents the results of successful writes. It can be something as simple as `Void` for now, or number of affected rows, or whatever.

Reply:
I see. I we return a number of applied mutations, but I'd have to count/aggregate that number. Won't that affect the write performance?

Reply:
Then I suggest you use `PCollection<Void>` or better `PCollection<?>` for now as the successful result type. Then you don't need to aggregate anything - but, if used as an input to Wait.on(), it will be "aggregated" very cheaply (all values ignored).

Reply:
Updated.